common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -505,74 +525,402 @@ private ImmutableSetMultimap<Class<? extends Annotation>, Element> toClassKeyedM
+   * alone is not sufficient to make distinction in all overloaded cases. For example, {@code <C
+   * extends Set<String>> void m(C c) {}}, and {@code <C extends SortedSet<String>> void m(C c) {}}
+   * both have the same toString {@code <C>m(C)} but are valid cases for overloading methods.
+   * Moreover, the needed enclosing type-element information is not included in the toString.
+   *
+   * <p>The executable element is retrieved by saving its enclosing type element, simple name, and
+   * ordinal position in the source code relative to other related overloaded methods, meaning those
+   * with the same simple name. This is possible because according to Java language specification
+   * for {@link TypeElement#getEnclosedElements()}: "As a particular instance of the general
+   * accuracy requirements and the ordering behavior required of this interface, the list of
+   * enclosed elements will be returned to the natural order for the originating source of
+   * information about the type. For example, if the information about the type is originating from
+   * a source file, the elements will be returned in source code order. (However, in that case the
+   * ordering of implicitly declared elements, such as default constructors, is not specified.)"
    *
-   * <p>Packages declared (and annotated) in {@code package-info.java} are tracked as deferred
-   * packages, type elements are tracked directly, and all other elements are tracked via their
-   * nearest enclosing type.
+   * <p>Simple name is saved since comparing the toString as a subtitle is not reliable when at
+   * least one parameter references ERROR, possibly because it is not generated yet. For example,
+   * method {@code void m(SomeGeneratedClass sgc)}, before the generation of {@code
+   * SomeGeneratedClass} has the toString {@code m(SomeGeneratedClass)}; however, after the
+   * generation it will have toString equal to {@code m(test.SomeGeneratedClass)} assuming that the
+   * package name is "test".
    */
-  private static final class ElementName {
-    private enum Kind {
-      PACKAGE_NAME,
-      TYPE_NAME,
+  private static final class ExecutableElementBluePrint extends ElementBluePrint {
+    private final TypeElementBluePrint enclosingTypeElementBluePrint;
+    private final Name simpleName;
+    private final int ordinalOverloadPosition;
+
+    private ExecutableElementBluePrint(Element element) {
+      super(element);
+      TypeElement enclosingTypeElement = getEnclosingType(element);
+      this.enclosingTypeElementBluePrint = new TypeElementBluePrint(enclosingTypeElement);
+      this.simpleName = element.getSimpleName();
+      int ordinalOverloadPosition = 1;
+      Iterator<? extends Element> iter = enclosingTypeElement.getEnclosedElements().iterator();
+      Element elt = iter.next(); // The collection is not empty
+      while (!element.equals(elt)) {
+        if (elt.getSimpleName().equals(element.getSimpleName()) && isAcceptableElementKind(elt)) {
```

Comment:
I think this should maybe be `isAcceptableElementKind(element)`? Otherwise the check should be outside the loop.

Reply:
I think mine is correct. Could you please double-check? The `elt` is the changing variable.

Reply:
I have changed `!element.equals(elt)` to `!elt.equals(element)` as I think that was the source of confusion.

Reply:
Yeah, I was the one getting confused here.